packages/components/src/modal/use-modal-exit-animation.ts

```diff
@@ -0,0 +1,75 @@
+/**
+ * WordPress dependencies
+ */
+import { useReducedMotion } from '@wordpress/compose';
+import { useCallback, useRef, useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { CONFIG } from '../utils';
+
+// Animation duration (ms) extracted to JS in order to be used on a setTimeout.
+const FRAME_ANIMATION_DURATION = CONFIG.transitionDuration;
+const FRAME_ANIMATION_DURATION_NUMBER = Number.parseInt(
+	CONFIG.transitionDuration
+);
+
+const EXIT_ANIMATION_NAME = 'components-modal__disappear-animation';
+
+export function useModalExitAnimation() {
+	const frameRef = useRef< HTMLDivElement >();
+	const [ isAnimatingOut, setIsAnimatingOut ] = useState( false );
+	const isReducedMotion = useReducedMotion();
+
+	const closeModal = useCallback(
+		( onAnimationEnd: () => void ) => {
+			function handleAnimationEnd( e: AnimationEvent ) {
+				if ( e.animationName === EXIT_ANIMATION_NAME ) {
+					setIsAnimatingOut( false );
+					onAnimationEnd();
+				}
+			}
+
+			if ( isReducedMotion ) {
+				onAnimationEnd();
+				return;
+			}
+
+			if ( ! frameRef.current ) {
+				return;
+			}
+
+			// Grab a reference to the frame element, to make sure we're referencing
+			// the same reference in the cleanup function.
+			const frameEl = frameRef.current;
+
+			setIsAnimatingOut( true );
+			frameEl.addEventListener( 'animationend', handleAnimationEnd, {
+				once: true,
+			} );
+			const animationTimeout = window.setTimeout( () => {
+				setIsAnimatingOut( false );
+				onAnimationEnd();
+			}, FRAME_ANIMATION_DURATION_NUMBER );
```

Comment:
There should be a "race" (maybe even a literal `Promise.race`) between the `animationend` listener and the timeout. Whoever wins will reset the state and call `onAnimationEnd`.

With the current code `onAnimationEnd` will be often called twice.

Reply:
Good point. I refactored to `Promise.race` as part of https://github.com/WordPress/gutenberg/pull/65203/commits/0ab921dadfde8e40c7a69714b1ded43d88f2c701, which should address this comment.